In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import os

from sklearn.ensemble import IsolationForest
from sklearn.model_selection import train_test_split
from sklearn import preprocessing

from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

In [2]:
df = pd.read_csv(os.getcwd() + "/df_all.csv")
df

peak_amplitude  integral  phase_angle  cycle_number  rise_time  \
0              0.956876     -4896          163         32698          0   
1              3.558275     -9372          152         31507          0   
2             16.040210     20326          113         35611          0   
3             17.296037     18912          221         32734          0   
4              6.600233    -17240          126         28314          0   
...                 ...       ...          ...           ...        ...   
5286175       12.029138      3151          275         15542          0   
5286176        9.960956    -17648          118         38209          0   
5286177        2.715035      2542          158         22870          0   
5286178        3.945221      3221          127         24506          0   
5286179        1.812937      1227           36         20297          1   

         pulse_width type  
0                  0  neg  
1                  0  pos  
2                 32  neg  
3                  0  neg  
4                  0  pos  
...              ...  ...  
5286175            0  pos  
5286176            0  pos  
5286177            0  neg  
5286178            0  neg  
5286179           17  pos  

[5286180 rows x 7 columns]

In [3]:
le = preprocessing.LabelEncoder()
X = df.iloc[:,:-1].values
y = le.fit_transform(df.iloc[:,-1])

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [5]:
isoforest = IsolationForest(n_estimators = 100)
isoforest.fit(X_train)

IsolationForest()

In [6]:
y_pred =  isoforest.predict(X_test)

In [7]:
pred_data = pd.DataFrame({"Predicted values":y_pred, "Actual values":y_test})
pred_data

Predicted values  Actual values
0                       1              0
1                       1              1
2                       1              0
3                       1              0
4                       1              0
...                   ...            ...
1057231                 1              0
1057232                 1              0
1057233                 1              0
1057234                 1              0
1057235                 1              0

[1057236 rows x 2 columns]

In [8]:
cm= confusion_matrix(y_test,y_pred)
print(cm)

[[     0      0      0]
 [128071      0 490798]
 [ 62257      0 376110]]


## y_pred
Isolation Tree
pos = -1
neg = 1

## y_test
Label Encoder
pos = 1
neg = 0

In [9]:
count = 0 
for i in range(len(y_test)):
    if (y_pred[i] == -1 and y_test[i] == 1):
        count += 1
    elif(y_pred[i] == 1 and y_test[i] == 0):
        count +=1
print(count)
print(count/len(y_test))

553055
0.5231140445463454


In [10]:
for i in range(6):
    test = df
    test = test.drop(test.columns[i], axis=1)
    print(test.columns.values)
    le = preprocessing.LabelEncoder()

    X = test.iloc[:,:-1]
    y = le.fit_transform(test.iloc[:,-1])
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 0)
    
    isoforest = IsolationForest(n_estimators = 100)
    isoforest.fit(X_train)
    y_pred =  isoforest.predict(X_test)
    cm= confusion_matrix(y_test,y_pred)
    print(cm)
    count = 0 
    for i in range(len(y_test)):
        if (y_pred[i] == -1 and y_test[i] == 1):
            count += 1
        elif(y_pred[i] == 1 and y_test[i] == 0):
            count +=1
    print(count/len(y_test))

['integral' 'phase_angle' 'cycle_number' 'rise_time' 'pulse_width' 'type']
[[     0      0      0]
 [122365      0 496504]
 [ 57917      0 380450]]
0.5244060928685743
['peak_amplitude' 'phase_angle' 'cycle_number' 'rise_time' 'pulse_width'
 'type']
[[     0      0      0]
 [151904      0 466965]
 [ 72262      0 366105]]
0.5100346564059491
['peak_amplitude' 'integral' 'cycle_number' 'rise_time' 'pulse_width'
 'type']
[[     0      0      0]
 [134972      0 483897]
 [ 64600      0 373767]]
0.5188028027800794
['peak_amplitude' 'integral' 'phase_angle' 'rise_time' 'pulse_width'
 'type']
[[     0      0      0]
 [130275      0 488594]
 [ 68980      0 369387]]
0.527388397670908
['peak_amplitude' 'integral' 'phase_angle' 'cycle_number' 'pulse_width'
 'type']
[[     0      0      0]
 [162147      0 456722]
 [100768      0 337599]]
0.5273089452118543
['peak_amplitude' 'integral' 'phase_angle' 'cycle_number' 'rise_time'
 'type']
[[     0      0      0]
 [124991      0 493878]
 [ 69204      0 369

## Isolation Tree Iteration Result


Removed components: 0 (peak_amplitude) | Accuracy score: 0.5244060928685743
Removed components: 1 (integral)       | Accuracy score: 0.5100346564059491
Removed components: 2 (phase_angle)    | Accuracy score: 0.5188028027800794
Removed components: 3 (cycle_number)   | Accuracy score: 0.527388397670908
Removed components: 4 (rise_time)      | Accuracy score: 0.5273089452118543
Removed components: 5 (pulse_width)    | Accuracy score: 0.5325982089145659